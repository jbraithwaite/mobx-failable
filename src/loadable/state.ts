import {Enum} from 'typescript-string-enums';

// tslint:disable-next-line:variable-name
export const State = Enum({
  /**
   * Denotes the absence of data and no requests in flight. This state
   * naturally occurs only once in the lifecycle of a Loadable.
   */
  empty: 'empty',

  /**
   * Denotes the absence of data and a request in flight.
   */
  pending: 'pending',

  /**
   * Denotes the presence of a value and no requests in flight.
   */
  success: 'success',

  /**
   * Denotes the presence of a value and a request in flight.
   */
  reloading: 'reloading',

  /**
   * Denotes the presence of an error and no requests in flight.
   */
  failure: 'failure',

  /**
   * Denotes the presence of an error and a request in flight.
   */
  retrying: 'retrying',
});

/**
 * Each of the six Loadable states is composed of two traits: availability and
 * flight. Availability refers to if there is no data, some data, or some
 * error. Flight, also known as "loading", refers to whether there is an
 * ongoing request.
 */
export type State = Enum<typeof State>;
