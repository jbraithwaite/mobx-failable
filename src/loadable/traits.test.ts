import {Enum} from 'typescript-string-enums';

import {State} from './state';

import {
  Availability,
  availabilityOf,
  Flight,
  flightOf,
  toState,
  TraitSet,
  withAvailability,
  withFlight,
} from './traits';

describe('TraitSet', () => {
  const states = Enum.values(State);
  const traitSets: TraitSet[] = states.map(TraitSet);

  const flights = Enum.values(Flight);
  const availabilities = Enum.values(Availability);

  describe('constructor', () => {
    for (const state of states) {
      it(`returns correctly for state ${state}`, () => {
        const result = TraitSet(state);
        expect(result).toMatchSnapshot();
      });
    }
  });

  describe('toState', () => {
    for (const traits of traitSets) {
      it(`returns correctly for TraitSet ${traits}`, () => {
        const result = toState(traits);
        expect(result).toMatchSnapshot();
      });
    }
  });

  describe('withFlight', () => {
    for (const state of states) {
      for (const flight of flights) {
        it(`changes state ${state} to flight ${flight}`, () => {
          const result = withFlight(state, flight);
          expect(result).toMatchSnapshot();
        });
      }
    }
  });

  describe('flightOf', () => {
    describe('given states', () => {
      for (const state of states) {
        it(`returns the correct flight for state ${state}`, () => {
          const result = flightOf(state);
          expect(result).toMatchSnapshot();
        });
      }
    });

    describe('given trait sets', () => {
      for (const traits of traitSets) {
        it(`returns the correct flight for trait set ${traits}`, () => {
          const result = flightOf(traits);
          expect(result).toMatchSnapshot();
        });
      }
    });
  });

  describe('withAvailability', () => {
    for (const state of states) {
      for (const availability of availabilities) {
        it(`changes state ${state} to availability ${availability}`, () => {
          const result = withAvailability(state, availability);
          expect(result).toMatchSnapshot();
        });
      }
    }
  });

  describe('availabilityOf', () => {
    describe('given states', () => {
      for (const state of states) {
        it(`returns the correct flight for state ${state}`, () => {
          const result = availabilityOf(state);
          expect(result).toMatchSnapshot();
        });
      }
    });

    describe('given trait sets', () => {
      for (const traits of traitSets) {
        it(`returns the correct flight for trait set ${traits}`, () => {
          const result = availabilityOf(traits);
          expect(result).toMatchSnapshot();
        });
      }
    });
  });
});
