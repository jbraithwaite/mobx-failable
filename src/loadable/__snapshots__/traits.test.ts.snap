// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TraitSet availabilityOf given states returns the correct flight for state empty 1`] = `0`;

exports[`TraitSet availabilityOf given states returns the correct flight for state failure 1`] = `2`;

exports[`TraitSet availabilityOf given states returns the correct flight for state pending 1`] = `0`;

exports[`TraitSet availabilityOf given states returns the correct flight for state reloading 1`] = `1`;

exports[`TraitSet availabilityOf given states returns the correct flight for state retrying 1`] = `2`;

exports[`TraitSet availabilityOf given states returns the correct flight for state success 1`] = `1`;

exports[`TraitSet availabilityOf given trait sets returns the correct flight for trait set 0 1`] = `0`;

exports[`TraitSet availabilityOf given trait sets returns the correct flight for trait set 1 1`] = `0`;

exports[`TraitSet availabilityOf given trait sets returns the correct flight for trait set 2 1`] = `1`;

exports[`TraitSet availabilityOf given trait sets returns the correct flight for trait set 3 1`] = `1`;

exports[`TraitSet availabilityOf given trait sets returns the correct flight for trait set 4 1`] = `2`;

exports[`TraitSet availabilityOf given trait sets returns the correct flight for trait set 5 1`] = `2`;

exports[`TraitSet constructor returns correctly for state empty 1`] = `0`;

exports[`TraitSet constructor returns correctly for state failure 1`] = `4`;

exports[`TraitSet constructor returns correctly for state pending 1`] = `1`;

exports[`TraitSet constructor returns correctly for state reloading 1`] = `3`;

exports[`TraitSet constructor returns correctly for state retrying 1`] = `5`;

exports[`TraitSet constructor returns correctly for state success 1`] = `2`;

exports[`TraitSet flightOf given states returns the correct flight for state empty 1`] = `0`;

exports[`TraitSet flightOf given states returns the correct flight for state failure 1`] = `0`;

exports[`TraitSet flightOf given states returns the correct flight for state pending 1`] = `1`;

exports[`TraitSet flightOf given states returns the correct flight for state reloading 1`] = `1`;

exports[`TraitSet flightOf given states returns the correct flight for state retrying 1`] = `1`;

exports[`TraitSet flightOf given states returns the correct flight for state success 1`] = `0`;

exports[`TraitSet flightOf given trait sets returns the correct flight for trait set 0 1`] = `0`;

exports[`TraitSet flightOf given trait sets returns the correct flight for trait set 1 1`] = `1`;

exports[`TraitSet flightOf given trait sets returns the correct flight for trait set 2 1`] = `0`;

exports[`TraitSet flightOf given trait sets returns the correct flight for trait set 3 1`] = `1`;

exports[`TraitSet flightOf given trait sets returns the correct flight for trait set 4 1`] = `0`;

exports[`TraitSet flightOf given trait sets returns the correct flight for trait set 5 1`] = `1`;

exports[`TraitSet toState returns correctly for TraitSet 0 1`] = `"empty"`;

exports[`TraitSet toState returns correctly for TraitSet 1 1`] = `"pending"`;

exports[`TraitSet toState returns correctly for TraitSet 2 1`] = `"success"`;

exports[`TraitSet toState returns correctly for TraitSet 3 1`] = `"reloading"`;

exports[`TraitSet toState returns correctly for TraitSet 4 1`] = `"failure"`;

exports[`TraitSet toState returns correctly for TraitSet 5 1`] = `"retrying"`;

exports[`TraitSet withAvailability changes state empty to availability 0 1`] = `"empty"`;

exports[`TraitSet withAvailability changes state empty to availability 1 1`] = `"success"`;

exports[`TraitSet withAvailability changes state empty to availability 2 1`] = `"failure"`;

exports[`TraitSet withAvailability changes state empty to availability error 1`] = `"empty"`;

exports[`TraitSet withAvailability changes state empty to availability none 1`] = `"empty"`;

exports[`TraitSet withAvailability changes state empty to availability value 1`] = `"empty"`;

exports[`TraitSet withAvailability changes state failure to availability 0 1`] = `"empty"`;

exports[`TraitSet withAvailability changes state failure to availability 1 1`] = `"success"`;

exports[`TraitSet withAvailability changes state failure to availability 2 1`] = `"failure"`;

exports[`TraitSet withAvailability changes state failure to availability error 1`] = `"empty"`;

exports[`TraitSet withAvailability changes state failure to availability none 1`] = `"empty"`;

exports[`TraitSet withAvailability changes state failure to availability value 1`] = `"empty"`;

exports[`TraitSet withAvailability changes state pending to availability 0 1`] = `"pending"`;

exports[`TraitSet withAvailability changes state pending to availability 1 1`] = `"reloading"`;

exports[`TraitSet withAvailability changes state pending to availability 2 1`] = `"retrying"`;

exports[`TraitSet withAvailability changes state pending to availability error 1`] = `"pending"`;

exports[`TraitSet withAvailability changes state pending to availability none 1`] = `"pending"`;

exports[`TraitSet withAvailability changes state pending to availability value 1`] = `"pending"`;

exports[`TraitSet withAvailability changes state reloading to availability 0 1`] = `"pending"`;

exports[`TraitSet withAvailability changes state reloading to availability 1 1`] = `"reloading"`;

exports[`TraitSet withAvailability changes state reloading to availability 2 1`] = `"retrying"`;

exports[`TraitSet withAvailability changes state reloading to availability error 1`] = `"pending"`;

exports[`TraitSet withAvailability changes state reloading to availability none 1`] = `"pending"`;

exports[`TraitSet withAvailability changes state reloading to availability value 1`] = `"pending"`;

exports[`TraitSet withAvailability changes state retrying to availability 0 1`] = `"pending"`;

exports[`TraitSet withAvailability changes state retrying to availability 1 1`] = `"reloading"`;

exports[`TraitSet withAvailability changes state retrying to availability 2 1`] = `"retrying"`;

exports[`TraitSet withAvailability changes state retrying to availability error 1`] = `"pending"`;

exports[`TraitSet withAvailability changes state retrying to availability none 1`] = `"pending"`;

exports[`TraitSet withAvailability changes state retrying to availability value 1`] = `"pending"`;

exports[`TraitSet withAvailability changes state success to availability 0 1`] = `"empty"`;

exports[`TraitSet withAvailability changes state success to availability 1 1`] = `"success"`;

exports[`TraitSet withAvailability changes state success to availability 2 1`] = `"failure"`;

exports[`TraitSet withAvailability changes state success to availability error 1`] = `"empty"`;

exports[`TraitSet withAvailability changes state success to availability none 1`] = `"empty"`;

exports[`TraitSet withAvailability changes state success to availability value 1`] = `"empty"`;

exports[`TraitSet withFlight changes state empty to flight 0 1`] = `"empty"`;

exports[`TraitSet withFlight changes state empty to flight 1 1`] = `"pending"`;

exports[`TraitSet withFlight changes state empty to flight busy 1`] = `"empty"`;

exports[`TraitSet withFlight changes state empty to flight idle 1`] = `"empty"`;

exports[`TraitSet withFlight changes state failure to flight 0 1`] = `"failure"`;

exports[`TraitSet withFlight changes state failure to flight 1 1`] = `"retrying"`;

exports[`TraitSet withFlight changes state failure to flight busy 1`] = `"failure"`;

exports[`TraitSet withFlight changes state failure to flight idle 1`] = `"failure"`;

exports[`TraitSet withFlight changes state pending to flight 0 1`] = `"empty"`;

exports[`TraitSet withFlight changes state pending to flight 1 1`] = `"pending"`;

exports[`TraitSet withFlight changes state pending to flight busy 1`] = `"empty"`;

exports[`TraitSet withFlight changes state pending to flight idle 1`] = `"empty"`;

exports[`TraitSet withFlight changes state reloading to flight 0 1`] = `"success"`;

exports[`TraitSet withFlight changes state reloading to flight 1 1`] = `"reloading"`;

exports[`TraitSet withFlight changes state reloading to flight busy 1`] = `"success"`;

exports[`TraitSet withFlight changes state reloading to flight idle 1`] = `"success"`;

exports[`TraitSet withFlight changes state retrying to flight 0 1`] = `"failure"`;

exports[`TraitSet withFlight changes state retrying to flight 1 1`] = `"retrying"`;

exports[`TraitSet withFlight changes state retrying to flight busy 1`] = `"failure"`;

exports[`TraitSet withFlight changes state retrying to flight idle 1`] = `"failure"`;

exports[`TraitSet withFlight changes state success to flight 0 1`] = `"success"`;

exports[`TraitSet withFlight changes state success to flight 1 1`] = `"reloading"`;

exports[`TraitSet withFlight changes state success to flight busy 1`] = `"success"`;

exports[`TraitSet withFlight changes state success to flight idle 1`] = `"success"`;
